File: static/java/AirtableException.java
// ==========================================
// MyAirtable static runtime.
// ==========================================

package myairtable;

import java.util.Objects;

/**
 * Errors raised by the MyAirtable runtime.
 *
 * <p>A sealed, UNCHECKED exception hierarchy (java-plan-v2 §2.3.7): catch {@link AirtableException}
 * for everything, or a specific subclass — or pattern-match with {@code switch} — for targeted
 * handling. Unchecked because checked exceptions would poison every generated signature.
 */
public abstract sealed class AirtableException extends RuntimeException
    permits AirtableException.Http,
        AirtableException.Api,
        AirtableException.Decoding,
        AirtableException.InvalidUrl,
        AirtableException.MissingCredentials,
        AirtableException.Network,
        AirtableException.RateLimited {

  private AirtableException(String message, Throwable cause) {
    super(message, cause);
  }

  /** Non-2xx HTTP response. {@code body} is the raw response body if decodable as UTF-8. */
  public static final class Http extends AirtableException {
    private final int statusCode;
    private final String body;

    public Http(int statusCode, String body) {
      super(
          body == null || body.isEmpty()
              ? "AirtableException.Http(" + statusCode + ")"
              : "AirtableException.Http(" + statusCode + "): " + body,
          null);
      this.statusCode = statusCode;
      this.body = body;
    }

    public int statusCode() {
      return statusCode;
    }

    public String body() {
      return body;
    }

    @Override
    public boolean equals(Object other) {
      return other instanceof Http h && h.statusCode == statusCode && Objects.equals(h.body, body);
    }

    @Override
    public int hashCode() {
      return 31 * statusCode + Objects.hashCode(body);
    }
  }

  /**
   * Airtable returned a structured error envelope ({@code {"error": {"type": ..., "message":
   * ...}}}).
   */
  public static final class Api extends AirtableException {
    private final String code;
    private final String apiMessage;

    public Api(String code, String apiMessage) {
      super("AirtableException.Api(" + code + "): " + apiMessage, null);
      this.code = code;
      this.apiMessage = apiMessage;
    }

    public String code() {
      return code;
    }

    public String apiMessage() {
      return apiMessage;
    }

    @Override
    public boolean equals(Object other) {
      return other instanceof Api a && a.code.equals(code) && a.apiMessage.equals(apiMessage);
    }

    @Override
    public int hashCode() {
      return 31 * code.hashCode() + apiMessage.hashCode();
    }
  }

  /** JSON decoding failed. {@code cause} originates from Jackson. */
  public static final class Decoding extends AirtableException {
    public Decoding(Throwable cause) {
      super("AirtableException.Decoding: " + (cause == null ? null : cause.getMessage()), cause);
    }

    // Underlying decode errors are not comparable; treat all decoding errors as equal.
    @Override
    public boolean equals(Object other) {
      return other instanceof Decoding;
    }

    @Override
    public int hashCode() {
      return getClass().hashCode();
    }
  }

  /** A request URL could not be constructed. */
  public static final class InvalidUrl extends AirtableException {
    public InvalidUrl() {
      super("AirtableException.InvalidUrl", null);
    }

    @Override
    public boolean equals(Object other) {
      return other instanceof InvalidUrl;
    }

    @Override
    public int hashCode() {
      return getClass().hashCode();
    }
  }

  /** {@link AirtableClient} was initialized without an API key or base ID. */
  public static final class MissingCredentials extends AirtableException {
    public MissingCredentials() {
      super("AirtableException.MissingCredentials", null);
    }

    @Override
    public boolean equals(Object other) {
      return other instanceof MissingCredentials;
    }

    @Override
    public int hashCode() {
      return getClass().hashCode();
    }
  }

  /**
   * A transport-level failure (connect, DNS, socket, request timeout) — the request never produced
   * an HTTP response. {@code cause} is the underlying I/O exception. Thread interruption is
   * re-asserted before wrapping.
   */
  public static final class Network extends AirtableException {
    public Network(Throwable cause) {
      super("AirtableException.Network: " + (cause == null ? null : cause.getMessage()), cause);
    }

    // Underlying transport errors are not comparable; treat all network errors as equal.
    @Override
    public boolean equals(Object other) {
      return other instanceof Network;
    }

    @Override
    public int hashCode() {
      return getClass().hashCode();
    }
  }

  /**
   * Airtable returned 429 Too Many Requests. {@code retryAfterSeconds} is the {@code Retry-After}
   * header value in seconds if present (else {@code null}).
   */
  public static final class RateLimited extends AirtableException {
    private final Double retryAfterSeconds;

    public RateLimited(Double retryAfterSeconds) {
      super(
          retryAfterSeconds != null
              ? "AirtableException.RateLimited(retryAfter: " + retryAfterSeconds + "s)"
              : "AirtableException.RateLimited",
          null);
      this.retryAfterSeconds = retryAfterSeconds;
    }

    public Double retryAfterSeconds() {
      return retryAfterSeconds;
    }

    @Override
    public boolean equals(Object other) {
      return other instanceof RateLimited r
          && Objects.equals(r.retryAfterSeconds, retryAfterSeconds);
    }

    @Override
    public int hashCode() {
      return Objects.hashCode(retryAfterSeconds);
    }
  }
}

File: static/kotlin/AirtableException.kt
// ==========================================
// MyAirtable static runtime.
// ==========================================

package myairtable

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonPrimitive

/**
 * Errors raised by the MyAirtable runtime.
 *
 * A sealed exception hierarchy: catch [AirtableException] for everything, or a
 * specific subclass for targeted handling.
 */
sealed class AirtableException(
    message: String,
    cause: Throwable? = null,
) : Exception(message, cause) {
    /** Non-2xx HTTP response. [body] is the raw response body if decodable as UTF-8. */
    class Http(
        val statusCode: Int,
        val body: String?,
    ) : AirtableException(if (body.isNullOrEmpty()) "AirtableException.Http($statusCode)" else "AirtableException.Http($statusCode): $body") {
        override fun equals(other: Any?): Boolean = other is Http && other.statusCode == statusCode && other.body == body

        override fun hashCode(): Int = 31 * statusCode + (body?.hashCode() ?: 0)
    }

    /** Airtable returned a structured error envelope (`{"error": {"type": ..., "message": ...}}`). */
    class Api(
        val code: String,
        val apiMessage: String,
    ) : AirtableException("AirtableException.Api($code): $apiMessage") {
        override fun equals(other: Any?): Boolean = other is Api && other.code == code && other.apiMessage == apiMessage

        override fun hashCode(): Int = 31 * code.hashCode() + apiMessage.hashCode()
    }

    /** JSON decoding failed. [cause] originates from kotlinx.serialization. */
    class Decoding(
        cause: Throwable,
    ) : AirtableException("AirtableException.Decoding: ${cause.message}", cause) {
        // Underlying decode errors are not comparable; treat all decoding errors as equal.
        override fun equals(other: Any?): Boolean = other is Decoding

        override fun hashCode(): Int = javaClass.hashCode()
    }

    /** A request URL could not be constructed. */
    class InvalidUrl : AirtableException("AirtableException.InvalidUrl") {
        override fun equals(other: Any?): Boolean = other is InvalidUrl

        override fun hashCode(): Int = javaClass.hashCode()
    }

    /** [AirtableClient] was initialized without an API key or base ID. */
    class MissingCredentials : AirtableException("AirtableException.MissingCredentials") {
        override fun equals(other: Any?): Boolean = other is MissingCredentials

        override fun hashCode(): Int = javaClass.hashCode()
    }

    /**
     * A transport-level failure (connect, DNS, socket, request timeout) — the
     * request never produced an HTTP response. [cause] is the underlying I/O
     * or timeout exception. Coroutine cancellation is never wrapped here.
     */
    class Network(
        cause: Throwable,
    ) : AirtableException("AirtableException.Network: ${cause.message}", cause) {
        // Underlying transport errors are not comparable; treat all network errors as equal.
        override fun equals(other: Any?): Boolean = other is Network

        override fun hashCode(): Int = javaClass.hashCode()
    }

    /**
     * Airtable returned 429 Too Many Requests. [retryAfterSeconds] is the
     * `Retry-After` header value in seconds if present.
     */
    class RateLimited(
        val retryAfterSeconds: Double?,
    ) : AirtableException(
            if (retryAfterSeconds != null) "AirtableException.RateLimited(retryAfter: ${retryAfterSeconds}s)" else "AirtableException.RateLimited",
        ) {
        override fun equals(other: Any?): Boolean = other is RateLimited && other.retryAfterSeconds == retryAfterSeconds

        override fun hashCode(): Int = retryAfterSeconds?.hashCode() ?: 0
    }
}

/**
 * Wire-format Airtable error envelope. Matches the shape Airtable returns on
 * non-2xx responses: `{"error": {"type": "...", "message": "..."}}` or
 * `{"error": "STRING"}` for some legacy endpoints.
 */
@Serializable
internal data class AirtableErrorEnvelope(
    val error: JsonElement,
) {
    /** Convert to the public exception type. */
    fun toAirtableException(): AirtableException =
        when (error) {
            is JsonObject -> {
                val type = (error["type"] as? JsonPrimitive)?.contentOrNull ?: "UNKNOWN"
                val message = (error["message"] as? JsonPrimitive)?.contentOrNull ?: ""
                AirtableException.Api(code = type, apiMessage = message)
            }

            is JsonPrimitive -> {
                AirtableException.Api(code = "UNKNOWN", apiMessage = error.jsonPrimitive.content)
            }

            else -> {
                AirtableException.Api(code = "UNKNOWN", apiMessage = error.toString())
            }
        }
}
